### PR TITLE
Fix redundant assignment removal in combination with break / continue.

### LINKED ISSUE #8072: [yul] optimizer's redundant assignment eliminator incorrectly removes non-redundant assignments in for loops containing a break/continue statement inside a nested branch statement
## Description

Found by LPM + libFuzzer + custom yul mutator (see #7947 )

```
{
        for {let i:= 0} lt(i,2) {i := add(i,1)}
        {
                // x is declared and implicitly
                // initialized to zero.
                let x
                // This assignment is not redundant
                // since it is used by mstore statement
                // that follows the if statement
                x := 1337
                if lt(i,1) {
                        // This assignment is redundant
                        x := 42
                        continue
                }
                mstore(0, x)
        }
}
```

is incorrectly optimized to

```
        {
            let i := 0
            for { } lt(i, 2) { i := add(i, 1) }
            {
                if lt(i, 1) { continue }
                mstore(0, 0)
            }
        }
```

`mstore(0, 0)` should be `mstore(0, 1337)` instead.

In the example above, the `if` statement could be replaced by another branching statement e.g., `switch` statement with the same result.

In general, this problem manifests while the redundant assignment eliminator is visiting a continue statement inside a branch inside a for loop: Assignments to a variable in the path leading to the branching statement are also removed apart from the correct redundant assignment within the branching statement.

For example, the yul optimizer also incorrectly marks (and subsequently removes) the assignment statement directly in the for loop below leading to `mstore(0, 1337)` although the correct optimization should be `mstore(0, 31337)`

```
{
        for {let i:= 0} lt(i,2) {i := add(i,1)}
        {
                let x := 1337
                // This assignment is incorrectly removed
                x := 31337
                if lt(i,1) {
                        // Redundant assignment eliminator removes all
                        // assignments to `x` in this path (including `x := 31337`)
                        // when it should be removing only `x := 42`
                        x := 42
                        continue
                }
                mstore(0, x)
        }
}
```

Unfortunately, this affects `v0.6.0` and may warrant a bug list entry. I'm trying to understand why this was not found earlier. My intuition is that the custom mutator had a role to play by increasing the likelihood (during fuzzing) of insertion of `continue` statement inside for loops (which I undertook only recently).

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc --strict-assembly --optimize <test.yul>
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/8072

### CODE REVIEW

**Comment by chriseth** [comment]
Changed the bug list entry only to say "introduced: 0.6.0", because that is when the yul optimizer was activated. There is another PR that fixes it and adds another bug list entry which states "low" severity before 0.6.0.

File: docs/bugs.json [hunk 1]
```diff
@@ -1,4 +1,15 @@
 [
+    {
+        "name": "YulOptimizerRedundantAssignmentBreakContinue",
+        "summary": "The Yul optimizer can remove essential assignments to variables declared inside for loops when Yul's continue or break statement is used. You are unlikely to be affected if you do not use inline assembly with for loops and continue and break statements.",
+        "description": "The Yul optimizer has a stage that remove assignments to variables that are overwritten again or are not used in all following control-flow branches. This logic incorrectly removes such assignments to variables declared inside a for loop if they can be removed in a control-flow branch that ends with ``break`` or ``continue`` even though they cannot be removed in other control-flow branches. Variables declared outside of the respective for loop are not affected.",
```
Comment (bshastry) [inline docs/bugs.json]:
```suggestion
        "description": "The Yul optimizer has a stage that removes assignments to variables that are overwritten before use or are not used in all following control-flow branches. This logic incorrectly removes such assignments to variables declared inside a for loop if they can be removed in a control-flow branch that ends with ``break`` or ``continue`` even though they cannot be removed in other control-flow branches. Variables declared outside of the respective for loop are not affected.",
```

**Review by bshastry — COMMENTED** [review]
Minor wording suggestion. Will fuzz the fix overnight and review the code tomorrow.

**Comment by bshastry** [comment]
Code and tests look good to me. Fuzzing overnight did not find anything. I'm trying to check if the bug may be introduced by var decl/assignments  created by function inliner as you mentioned.

`osx` build seems to fail because of a compiler warning -> error (unused variable in SolidityEndToEndTests.cpp). Seems unrelated to this PR.

**Comment by leonardoalt** [comment]
Build failure

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -14319,8 +14319,6 @@ BOOST_AUTO_TEST_CASE(event_wrong_abi_name)
 	)";
 	compileAndRun(sourceCode, 0, "ClientReceipt", bytes());
 	compileAndRun(sourceCode, 0, "Test", bytes(), map<string, Address>{{"ClientReceipt", m_contractAddress}});
-	u256 value(18);
-	u256 id(0x1234);
```
Comment (leonardoalt) [inline test/libsolidity/SolidityEndToEndTest.cpp:14323]: Not sure why this didn't break builds before

**Pull request merged**